File: 15. 3Sum.py
class Solution:
    def threeSum(self, nums):
        res = []
        nums.sort()
        # 由於需要三個數字，用two pointer會少一個 所以一定要先跑一次Loop
        for i in range(len(nums)-2):
            # 由於排序過，只要當下的數字跟前一個一樣，就跳過以免重複
            if i > 0 and nums[i] == nums[i-1]:
                continue
            # 左右抓好
            l, r = i+1, len(nums)-1
            # Two pointer
            while l < r:
                # 三個值有沒有等於0
                s = nums[i] + nums[l] + nums[r]
                # 如果小就往右加
                if s < 0:
                    l +=1
                # 如果大就往左減 
                elif s > 0:
                    r -= 1
                # 等於0加起來
                else:
                    res.append((nums[i], nums[l], nums[r]))
                    #需符合two-pointer條件，並且如果下一個數字一樣我就往後加避免重複 
                    while l < r and nums[l] == nums[l+1]:
                        l += 1
                    while l < r and nums[r] == nums[r-1]:
                        r -= 1
                    l += 1; r -= 1
        return res

File: C++/一知半解/Medium/15. 3Sum.cpp
#include<iostream>
#include <bits/stdc++.h>
#include <map>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> answer;
        //算總和都先排序
        sort(nums.begin(),nums.end());
        //由於題目要求在vector塞入不重複的組合但相加等於0的vector，因此要去想如何不會重複
        //最直觀的方法是，因為排序過了，只要我們迴圈發現現在跟前一個一樣，就跳過以免重複
        //並且left的值默認從現在的index往後，這樣就不會加到重複Case
        for(int index=0;index<nums.size();index++){
            if(index>0 && nums[index]==nums[index-1]){//排序過了，只要我們迴圈發現現在跟前一個一樣，就跳過以免重複
                continue;
            }
            int left=index+1;int right=nums.size()-1;//排序過了，只要left從現在的位置往右數，這樣就不會產生包含之前位置的組合，因為我們的算法會包含所有該數字的case，
            while(right>left){
                int target = nums[index]+nums[left]+nums[right];
                if(target<0){
                    left++;
                }
                else if(target>0){
                    right--;
                }
                else{
                    vector<int> zeroCase{nums[index],nums[left],nums[right]};
                    answer.push_back(zeroCase);
                    while(right>left && nums[left]==nums[left+1]){//一樣為了避免把重複case放進來，發現左邊一樣就可以一直往前走，後面迴圈碰到也會跳過
                        left++;
                    }   
                    while(right>left && nums[right]==nums[right-1]){//一樣為了避免把重複case放進來，發現右邊一樣就可以一直往回走，後面迴圈碰到也會跳過
                        right--;
                    }
                    left++;
                    right--;
                }
            }
        }
        return answer ;        
    }
};